lib/ProductOpener/APIProductServices.pm

```diff
@@ -178,57 +208,82 @@ Process API v3 product services requests.
-		# if the fields parameter value is "updated"
-		$request_ref->{updated_product_fields} = {};
-
-		foreach my $service (@{$request_body_ref->{services}}) {
-			my $service_function = $service_functions{$service};
-			if (defined $service_function) {
-				&$service_function($product_ref, $request_ref->{updated_product_fields});
-			}
-			else {
-				add_error(
-					$response_ref,
-					{
-						message => {id => "unknown_service"},
-						field => {id => "services", value => $service},
-						impact => {id => "failure"},
-					}
-				);
-			}
-		}
-
-		# Select / compute only the fields requested by the caller, default to updated fields
-		my $fields_ref = request_param($request_ref, 'fields');
-		if (not defined $fields_ref) {
-			$fields_ref = ["updated"];
-		}
-		$log->debug("product_services_api - before customize", {fields_ref => $fields_ref, product_ref => $product_ref})
-			if $log->is_debug();
-		$response_ref->{product} = customize_response_for_product($request_ref, $product_ref, undef, $fields_ref);
-
-		# Echo back the services that were executed
-		$response_ref->{fields} = $fields_ref;
-	}
-
-	$log->debug("product_services_api - stop", {request => $request_ref}) if $log->is_debug();
+    $log->debug("product_services_api - start", {request => $request_ref}) if $log->is_debug();
+
+    my $response_ref = $request_ref->{api_response};
+    my $request_body_ref = $request_ref->{body_json};
+
+    $log->debug("product_services_api - body", {request_body => $request_body_ref}) if $log->is_debug();
+
+    my $error = check_product_services_api_input($request_ref);
+
+    # If we did not get a fatal error, we can execute the services on the input product object
+    if (not $error) {
+
+        my $product_ref = $request_body_ref->{product};
+        $request_ref->{updated_product_fields} = {};
+
+        foreach my $service (@{$request_body_ref->{services}}) {
+            my $service_function = $service_functions{$service};
+
+            if (defined $service_function) {
+                if ($service eq 'check_quality') {
+                    # Create a temporary product reference for quality checks
+                    my $temp_product_ref = {};
+                    $temp_product_ref->{nutrition} = $request_body_ref->{nutrition} if defined $request_body_ref->{nutrition};
+                    $temp_product_ref->{ingredients} = $request_body_ref->{ingredients} if defined $request_body_ref->{ingredients};
+
+                    # Call the check_quality service, passing the temporary product ref
+                    &$service_function($temp_product_ref, $request_ref->{updated_product_fields});
+                    
+                    # Integrate the quality check results back into the main product_ref
+                    $product_ref->{data_quality_tags} = $temp_product_ref->{data_quality_tags} if defined $temp_product_ref->{data_quality_tags};
+                } else {
```

Comment:
The idea of this pattern is to avoid having specific code according to the service…

Also we will put the ingredients and nutrition fields under the "product" field so you have product_ref working.